src/rtc.c

```diff
@@ -0,0 +1,518 @@
+ * Use #JOYBUS_RTC_CONTROL_MODE_RUN after writing to block 2 to resume the RTC.
+ *
+ * It may take up to 20 milliseconds after this function returns for the
+ * write to actually complete on real hardware. It is recommended to wait
+ * using #JOYBUS_RTC_WRITE_DELAY after setting the RTC control block before
+ * calling #joybus_rtc_read_time or #joybus_rtc_write_time to ensure the
+ * control block data was fully written.
+ *
+ * It is strongly recommended to propagate the calibration data returned from
+ * #joybus_rtc_read_control into the calibration parameter of this function.
+ * If you have any reason to modify or omit the calibration data, please
+ * update this documentation with an explanation of why and how you would
+ * want to do this.
+ *
+ * Generally, you should not need to call this function directly. Prefer
+ * calling the high-level #joybus_rtc_run and #joybus_rtc_set functions which
+ * handle the delays and calibration propagation.
+ *
+ * @param[in]  control
+ *             The control block data
+ *
+ * @param[in]  calibration
+ *             The opaque calibration data from #joybus_rtc_read_control
+ */
+void joybus_rtc_write_control( uint16_t control, uint32_t calibration )
+{
+    uint64_t data = (uint64_t)control << 48 | calibration;
+    joybus_rtc_write( 0, &data );
+}
+
+/**
+ * @brief Write a new date/time to the real-time clock.
+ *
+ * This will fail silently if block 2 of the RTC is write-protected.
+ *
+ * The RTC control block should be set to #JOYBUS_RTC_CONTROL_MODE_SET
+ * using #joybus_rtc_write_control before calling this function.
+ *
+ * The RTC control block should be set to #JOYBUS_RTC_CONTROL_MODE_RUN
+ * using #joybus_rtc_write_control after calling this function.
+ *
+ * It may take up to 20 milliseconds after this function returns for the
+ * write to actually complete on real hardware. It is recommended to wait
+ * using #JOYBUS_RTC_WRITE_DELAY before setting the RTC control block back
+ * to #JOYBUS_RTC_CONTROL_MODE_RUN after setting the time.
+ *
+ * Generally, you should not need to call this function directly. Prefer
+ * calling the high-level #joybus_rtc_set function which handles the RTC
+ * control block writes, delays, and detects ignored writes.
+ *
+ * @param[in]  rtc_time
+ *             Source pointer for the RTC time data structure
+ *
+ * @return the status of the real-time clock
+ */
+void joybus_rtc_write_time( const rtc_time_t * rtc_time )
+{
+    uint8_t data[sizeof(uint64_t)];
+
+    int year = rtc_time->year - 1900;
+    data[0] = byte_to_bcd(rtc_time->sec);
+    data[1] = byte_to_bcd(rtc_time->min);
+    data[2] = byte_to_bcd(rtc_time->hour) + 0x80;
+    data[3] = byte_to_bcd(rtc_time->day);
+    data[4] = byte_to_bcd(rtc_time->week_day);
+    data[5] = byte_to_bcd(rtc_time->month + 1);
+    data[6] = byte_to_bcd(year);
+    data[7] = byte_to_bcd(year / 100);
+
+    joybus_rtc_write( 2, (uint64_t *)data );
```

Comment:
You can't cast a uint8_t array to uint64_t because you have no guarantees on the alignment.

Do the opposite: declare a uint64_t and then cast it to uint8_t* to access it byte-wise

Reply:
Addressed in https://github.com/DragonMinded/libdragon/pull/152/commits/e46fbb07a6cbb01a98f05a2d3b170d4ff1860b77